

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Machine-Learning on RNA-Seq data &mdash; master_project 0.0.1 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/gallery.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="python_scripts" href="modules.html" />
    <link rel="prev" title="Preprocessing" href="preprocessing_analysis.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> master_project
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="getting_started.html">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="metadata.html">Metadata acquisition</a></li>
<li class="toctree-l1"><a class="reference internal" href="preprocessing_analysis.html">Preprocessing</a></li>
<li class="toctree-l1"><a class="reference internal" href="preprocessing_analysis.html#analysis">Analysis</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Machine-Learning on RNA-Seq data</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#machine-learning">Machine Learning</a></li>
<li class="toctree-l2"><a class="reference internal" href="#kegg-database">KEGG database</a></li>
<li class="toctree-l2"><a class="reference internal" href="#feature-importances">Feature importances</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="modules.html">python_scripts</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">master_project</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Machine-Learning on RNA-Seq data</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/machine_learning.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="machine-learning-on-rna-seq-data">
<h1>Machine-Learning on RNA-Seq data<a class="headerlink" href="#machine-learning-on-rna-seq-data" title="Permalink to this headline">¶</a></h1>
<p>This chapter describes how to use the machine learning script and use different KEGG genes as filter.</p>
<div class="section" id="machine-learning">
<h2>Machine Learning<a class="headerlink" href="#machine-learning" title="Permalink to this headline">¶</a></h2>
<p>This script should be run within the docker container using the <code class="docutils literal notranslate"><span class="pre">python_scrips/environment.yml</span></code>.
For efficient investigation of different algorithms a wrapper bash command is outlined here.
It executes the machine_learning script for all implemented algorithms and writes STDOUT and STDERR to a log file.</p>
<p>Outputs:</p>
<ul class="simple">
<li><p>Plot of 10 most important features</p></li>
<li><p>Feature importances as csv file</p></li>
<li><p>Plot of confusion matrix</p></li>
<li><p>Plot of ROC-AUC curve</p></li>
</ul>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># Bash wrapper script</span>

<span class="nb">declare</span> -a <span class="nv">classifiers</span><span class="o">=(</span><span class="s2">&quot;LinearSVC&quot;</span> <span class="s2">&quot;SVC&quot;</span> <span class="s2">&quot;RandomForest&quot;</span> <span class="s2">&quot;MultiLayerPerceptron&quot;</span><span class="o">)</span>

<span class="k">for</span> val <span class="k">in</span> <span class="si">${</span><span class="nv">classifiers</span><span class="p">[@]</span><span class="si">}</span><span class="p">;</span> <span class="k">do</span>

    <span class="nb">echo</span> Starting to train <span class="nv">$val</span> classifier

    python container/app/machine_learning_tool.py <span class="se">\</span>
    -i/--inpath &lt;gene_counts/TPM.txt&gt; <span class="se">\</span>
    -m /--metadata &lt;metadata.csv&gt; <span class="se">\</span>
    -a/--algorithm <span class="nv">$val</span> <span class="se">\</span>
    <span class="o">[</span>-k/--kegg &lt;KEGG_filter.txt&gt;<span class="o">]</span> <span class="se">\</span>
    -o/--outpath &lt;destination directory&gt; <span class="se">\</span>
    -c/--cores &lt;INT&gt;  <span class="se">\</span>
    -t/-title &lt;tissue_dataset&gt; <span class="m">2</span>&gt;<span class="p">&amp;</span><span class="m">1</span> &gt;&gt; out.log

<span class="k">done</span>
</pre></div>
</div>
</div>
<div class="section" id="kegg-database">
<h2>KEGG database<a class="headerlink" href="#kegg-database" title="Permalink to this headline">¶</a></h2>
<p>For obtaining human or cancer genes from the KEGG database the following sources can be used.
The tables have to be merged in a database-like fashion to be input in the ML script.</p>
<ul class="simple">
<li><p>Source of human KEGG <a class="reference external" href="http://rest.kegg.jp/list/hsa">genes</a></p></li>
<li><p>Source of human KEGG <a class="reference external" href="http://rest.kegg.jp/list/pathway/hsa">pathways</a></p></li>
<li><p>Source of NCBI Entrez Identifiers to KEGG <a class="reference external" href="http://rest.kegg.jp/conv/hsa/ncbi-geneid">identifiers</a></p></li>
<li><p>Source of mapping NCBI Entrez to Ensembl <a class="reference external" href="https://www.genenames.org/cgi-bin/download/custom?col=gd_pub_eg_id&amp;col=gd_pub_ensembl_id&amp;status=Approved&amp;status=Entry%20Withdrawn&amp;hgnc_dbtag=on&amp;order_by=gd_app_sym_sort&amp;format=text&amp;submit=submit">ids</a></p></li>
</ul>
<ul class="simple">
<li><p>05200 <a class="reference external" href="https://www.kegg.jp/kegg-bin/show_pathway?hsa05200">Pathways in cancer</a> , (<a class="reference external" href="http://rest.kegg.jp/link/hsa/hsa05200">pathway list 5200</a>)</p></li>
<li><p>05202 <a class="reference external" href="https://www.kegg.jp/kegg-bin/show_pathway?hsa05202">Transcriptional misregulation in cancer</a> , (<a class="reference external" href="http://rest.kegg.jp/link/hsa/hsa05202">pathway list 5202</a>)</p></li>
<li><p>05206 <a class="reference external" href="https://www.kegg.jp/kegg-bin/show_pathway?hsa05206">MicroRNAs in cancer</a> , (<a class="reference external" href="http://rest.kegg.jp/link/hsa/hsa05206">pathway list 5206</a>)</p></li>
<li><p>05205 <a class="reference external" href="https://www.kegg.jp/kegg-bin/show_pathway?hsa05205">Proteoglycans in cancer</a> , (<a class="reference external" href="http://rest.kegg.jp/link/hsa/hsa05205">pathway list  5205</a>)</p></li>
<li><p>05204 <a class="reference external" href="https://www.kegg.jp/kegg-bin/show_pathway?hsa05204">Chemical carcinogenesis</a> , (<a class="reference external" href="http://rest.kegg.jp/link/hsa/hsa05204">pathway list 5204</a>)</p></li>
<li><p>05203 <a class="reference external" href="https://www.kegg.jp/kegg-bin/show_pathway?hsa05203">Viral carcinogenesis</a> , (<a class="reference external" href="http://rest.kegg.jp/link/hsa/hsa05203">pathway list 5203</a>)</p></li>
<li><p>05230 <a class="reference external" href="https://www.kegg.jp/kegg-bin/show_pathway?hsa05230">Central carbon metabolism in cancer</a> , (<a class="reference external" href="http://rest.kegg.jp/link/hsa/hsa05230">pathway list 5230</a>)</p></li>
<li><p>05231 <a class="reference external" href="https://www.kegg.jp/kegg-bin/show_pathway?hsa05231">Choline metabolism in cancer</a> , (<a class="reference external" href="http://rest.kegg.jp/link/hsa/hsa05231">pathway list 5231</a>)</p></li>
<li><p>05235 <a class="reference external" href="https://www.kegg.jp/kegg-bin/show_pathway?hsa05235">PD-L1 expression and PD-1 checkpoint pathway in cancer</a> , (<a class="reference external" href="http://rest.kegg.jp/link/hsa/hsa05235">pathway list 5235</a>)</p></li>
</ul>
<table class="colwidths-given docutils align-default" id="id1">
<caption><span class="caption-text">Excerpt of a KEGG_filter.txt file</span><a class="headerlink" href="#id1" title="Permalink to this table">¶</a></caption>
<colgroup>
<col style="width: 20%" />
<col style="width: 20%" />
<col style="width: 60%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>ensembl_id</p></th>
<th class="head"><p>GeneName</p></th>
<th class="head"><p>pathway</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>ENSG00000000419</p></td>
<td><p>DPM1</p></td>
<td><p>[‘path:hsa01100’, ‘path:hsa00510’]</p></td>
</tr>
<tr class="row-odd"><td><p>ENSG00000000938</p></td>
<td><p>FGR</p></td>
<td><p>[‘path:hsa04062’]</p></td>
</tr>
<tr class="row-even"><td><p>ENSG00000000971</p></td>
<td><p>CFH</p></td>
<td><p>[‘path:hsa04610’, ‘path:hsa05150’]</p></td>
</tr>
<tr class="row-odd"><td><p>ENSG00000001036</p></td>
<td><p>FUCA2</p></td>
<td><p>[‘path:hsa04142’, ‘path:hsa00511’]</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="feature-importances">
<h2>Feature importances<a class="headerlink" href="#feature-importances" title="Permalink to this headline">¶</a></h2>
<p>Annotation of cancer prognostics can be done with the <code class="docutils literal notranslate"><span class="pre">get_protein_atlas.py</span></code> script.
It deploys programmatic data access to the API of <a class="reference external" href="https://www.proteinatlas.org/">The Human Protein Atlas</a> and retrieves information in JSON format.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>get_protein_atlas.py
-i/--inpath:            Inpath to table with feature importance values <span class="o">(</span>CSV<span class="o">)</span>
-c/--cancer:            Cancer <span class="nb">type</span> used <span class="k">for</span> query
-o/--outpath:           Outpath <span class="k">for</span> annotated feature importance table <span class="o">(</span>CSV<span class="o">)</span>
--latex:                If set, print annotated table <span class="k">in</span> latex format to STDOUT

python get_protein_atlas.py -i &lt;feature_importance.csv&gt; -c &lt;Pancreatic/Liver&gt;  -o &lt;annotated_feature_importance.csv&gt;
</pre></div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="modules.html" class="btn btn-neutral float-right" title="python_scripts" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="preprocessing_analysis.html" class="btn btn-neutral float-left" title="Preprocessing" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021, Susanne Jodoin.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>