

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>fileUtils.file_handling &mdash; master_project 0.0.1 documentation</title>
  

  
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/gallery.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../index.html" class="icon icon-home"> master_project
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../getting_started.html">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../metadata.html">Metadata acquisition</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../preprocessing_analysis.html">Preprocessing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../preprocessing_analysis.html#analysis">Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../machine_learning.html">Machine-Learning on RNA-Seq data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../modules.html">python_scripts</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">master_project</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../index.html">Module code</a> &raquo;</li>
        
      <li>fileUtils.file_handling</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for fileUtils.file_handling</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">os.path</span>
<span class="kn">import</span> <span class="nn">glob</span>


<div class="viewcode-block" id="get_files"><a class="viewcode-back" href="../../fileUtils.html#fileUtils.file_handling.get_files">[docs]</a><span class="k">def</span> <span class="nf">get_files</span><span class="p">(</span><span class="n">inpath</span><span class="p">,</span> <span class="n">ext</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Create list of files</span>

<span class="sd">    :param inpath: to folder</span>
<span class="sd">    :param ext: of files</span>
<span class="sd">    :return: files: list</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">os</span><span class="o">.</span><span class="n">chdir</span><span class="p">(</span><span class="n">inpath</span><span class="p">)</span>
    <span class="n">files</span> <span class="o">=</span> <span class="p">[</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">f</span><span class="p">))</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="n">inpath</span> <span class="o">+</span> <span class="n">ext</span><span class="p">)]</span>
    <span class="k">return</span> <span class="n">files</span></div>


<div class="viewcode-block" id="get_files_and_sample_ids"><a class="viewcode-back" href="../../fileUtils.html#fileUtils.file_handling.get_files_and_sample_ids">[docs]</a><span class="k">def</span> <span class="nf">get_files_and_sample_ids</span><span class="p">(</span><span class="n">inpath</span><span class="p">,</span> <span class="n">ext</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Create list of files</span>

<span class="sd">    :param inpath: to files</span>
<span class="sd">    :param ext: file extension</span>
<span class="sd">    :return: files: as list</span>
<span class="sd">    :return: sample_ids: as list</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">os</span><span class="o">.</span><span class="n">chdir</span><span class="p">(</span><span class="n">inpath</span><span class="p">)</span>
    <span class="n">files</span> <span class="o">=</span> <span class="p">[</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">f</span><span class="p">))</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="n">inpath</span> <span class="o">+</span> <span class="n">ext</span><span class="p">)]</span>
    <span class="n">sample_ids</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">file</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">files</span><span class="p">):</span>
        <span class="n">file</span> <span class="o">=</span> <span class="n">file</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">file</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;PCAWG&#39;</span><span class="p">):</span>
            <span class="c1"># print(file.split(&quot;.&quot;)[1])</span>
            <span class="n">sample_ids</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">file</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># print(file.split(&quot;_&quot;)[0])</span>
            <span class="n">sample_ids</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">file</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>

    <span class="k">return</span> <span class="n">files</span><span class="p">,</span> <span class="n">sample_ids</span></div>


<div class="viewcode-block" id="read_unique_genes"><a class="viewcode-back" href="../../fileUtils.html#fileUtils.file_handling.read_unique_genes">[docs]</a><span class="k">def</span> <span class="nf">read_unique_genes</span><span class="p">(</span><span class="n">genes</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Read genes from rnaseq pipeline</span>

<span class="sd">    :param genes: genes_id mapped to name</span>
<span class="sd">    :return: gene_dict: mapping ids to names</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">gene_dict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">genes</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">f</span><span class="o">.</span><span class="n">readlines</span><span class="p">():</span>
            <span class="n">fields</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">gene_id</span> <span class="o">=</span> <span class="n">fields</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">gene_name</span> <span class="o">=</span> <span class="n">fields</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
            <span class="n">gene_dict</span><span class="p">[</span><span class="n">gene_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">gene_name</span>
    <span class="k">return</span> <span class="n">gene_dict</span></div>


<div class="viewcode-block" id="write_table"><a class="viewcode-back" href="../../fileUtils.html#fileUtils.file_handling.write_table">[docs]</a><span class="k">def</span> <span class="nf">write_table</span><span class="p">(</span><span class="n">gene_dict</span><span class="p">,</span> <span class="n">sample_ids</span><span class="p">,</span> <span class="n">all_files</span><span class="p">,</span> <span class="n">outpath</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Create tab separated table with merged values, adapted from Steffen Lemke</span>

<span class="sd">    :param gene_dict: gene_ids mapped to gene_names</span>
<span class="sd">    :param sample_ids: array</span>
<span class="sd">    :param all_files: dict</span>
<span class="sd">    :param outpath: path where to store outfile, tsv</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">outpath</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">table</span><span class="p">:</span>
        <span class="c1"># .join(list) iterates over ids</span>
        <span class="n">table</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;GeneID</span><span class="se">\t</span><span class="s2">GeneName</span><span class="se">\t</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">sample_ids</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="n">all_gene_ids</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">all_files</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
        <span class="n">all_gene_ids</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>

        <span class="k">for</span> <span class="n">gene_id</span> <span class="ow">in</span> <span class="n">all_gene_ids</span><span class="p">:</span>
            <span class="n">gene_name</span> <span class="o">=</span> <span class="n">gene_dict</span><span class="p">[</span><span class="n">gene_id</span><span class="p">]</span>
            <span class="n">table</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">gene_id</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="n">gene_name</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">tpm_values_gene_id_str</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">all_files</span><span class="p">[</span><span class="n">gene_id</span><span class="p">]]</span>
            <span class="n">table</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">tpm_values_gene_id_str</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="read_tpm"><a class="viewcode-back" href="../../fileUtils.html#fileUtils.file_handling.read_tpm">[docs]</a><span class="k">def</span> <span class="nf">read_tpm</span><span class="p">(</span><span class="n">infile</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Reads TPM values from merged file into floats</span>

<span class="sd">    :param infile: merged TPM values</span>
<span class="sd">    :return: sample_ids</span>
<span class="sd">    :return: gene_ids</span>
<span class="sd">    :return: gene_names</span>
<span class="sd">    :return: data_list</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">tmp</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">gene_names</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">gene_ids</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">header</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">infile</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">file</span><span class="p">:</span>

        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">file</span><span class="p">):</span>
            <span class="n">fields</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="c1"># first line: skip first two fields &quot;Gene ID&quot;, &quot;Gene Name&quot;</span>
                <span class="n">header</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">fields</span><span class="p">[</span><span class="mi">2</span><span class="p">:])</span>
            <span class="k">if</span> <span class="n">i</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="c1"># other lines: create lists</span>
                <span class="n">current_gene</span> <span class="o">=</span> <span class="n">fields</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">current_gene_name</span> <span class="o">=</span> <span class="n">fields</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
                <span class="n">tpm_values</span> <span class="o">=</span> <span class="n">fields</span><span class="p">[</span><span class="mi">2</span><span class="p">:]</span>
                <span class="c1"># in case that a gene has no tmp values --&gt; skip that line! (alternative: fill with zeros)</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">tpm_values</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="k">continue</span>
                <span class="n">gene_ids</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">current_gene</span><span class="p">)</span>
                <span class="n">gene_names</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">current_gene_name</span><span class="p">)</span>
                <span class="n">tmp</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">tpm_values</span><span class="p">)</span>

        <span class="c1"># header contains sample ids</span>
        <span class="n">sample_ids</span> <span class="o">=</span> <span class="p">[</span><span class="n">val</span> <span class="k">for</span> <span class="n">sublist</span> <span class="ow">in</span> <span class="n">header</span> <span class="k">for</span> <span class="n">val</span> <span class="ow">in</span> <span class="n">sublist</span><span class="p">]</span>

    <span class="n">file</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

    <span class="c1"># convert nested list elements to floats</span>
    <span class="n">data_list</span> <span class="o">=</span> <span class="p">[</span><span class="nb">list</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">float</span><span class="p">,</span> <span class="n">item</span><span class="p">))</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">tmp</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">sample_ids</span><span class="p">,</span> <span class="n">gene_ids</span><span class="p">,</span> <span class="n">gene_names</span><span class="p">,</span> <span class="n">data_list</span></div>


<div class="viewcode-block" id="parse_csv"><a class="viewcode-back" href="../../fileUtils.html#fileUtils.file_handling.parse_csv">[docs]</a><span class="k">def</span> <span class="nf">parse_csv</span><span class="p">(</span><span class="n">files</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Parse csv files</span>

<span class="sd">    :param files: list of files</span>
<span class="sd">    :return: srr_metadata: metadata dictionary</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">srr_metadata</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">is_tumor</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
    <span class="k">for</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">files</span><span class="p">:</span>
        <span class="c1"># Read each file</span>
        <span class="n">file</span> <span class="o">=</span> <span class="n">file</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">file</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="c1"># strip lines to not read empty  lines</span>
            <span class="n">lines</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">line</span> <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="p">(</span><span class="n">l</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">for</span> <span class="n">l</span> <span class="ow">in</span> <span class="n">f</span><span class="p">)</span> <span class="k">if</span> <span class="n">line</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">l</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">lines</span><span class="p">):</span>
                <span class="c1"># skip header</span>
                <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="k">continue</span>
                <span class="n">fields</span> <span class="o">=</span> <span class="n">l</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;,&quot;</span><span class="p">)</span>
                <span class="n">run</span> <span class="o">=</span> <span class="n">fields</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">bio_project</span> <span class="o">=</span> <span class="n">fields</span><span class="p">[</span><span class="mi">21</span><span class="p">]</span>
                <span class="n">bio_sample</span> <span class="o">=</span> <span class="n">fields</span><span class="p">[</span><span class="mi">25</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">fields</span><span class="p">[</span><span class="mi">36</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;no&#39;</span><span class="p">:</span>
                    <span class="n">is_tumor</span> <span class="o">=</span> <span class="s1">&#39;normal&#39;</span>
                <span class="k">if</span> <span class="n">fields</span><span class="p">[</span><span class="mi">36</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;yes&#39;</span><span class="p">:</span>
                    <span class="n">is_tumor</span> <span class="o">=</span> <span class="s1">&#39;tumor&#39;</span>
                <span class="k">if</span> <span class="n">run</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">srr_metadata</span><span class="p">:</span>
                    <span class="n">srr_metadata</span><span class="p">[</span><span class="n">run</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">bio_sample</span><span class="p">,</span> <span class="n">is_tumor</span><span class="p">,</span> <span class="n">bio_project</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">srr_metadata</span></div>
</pre></div>

           </div>
           
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021, Susanne Jodoin.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>